classes/check/tasklatencycheck.php

```diff
@@ -0,0 +1,139 @@
+     */
+    public function get_result(): result {
+        global $CFG, $DB;
+
+        $taskconfig = get_config('tool_heartbeat', 'tasklatencymonitoring');
+        $status = result::OK;
+        $messages = '';
+
+        $lockstats = isset($CFG->lock_factory) && $CFG->lock_factory === '\\tool_lockstats\\proxy_lock_factory';
+
+        $tasks = explode(PHP_EOL, $taskconfig);
+        foreach ($tasks as $taskconfig) {
+            list($taskclass, $runtime, $startdelay, $completiondelay) = explode(',', str_replace(' ', '', $taskconfig));
+            $task = \core\task\manager::get_scheduled_task($taskclass);
+
+            // Input validation.
+            $valid = $task !== false;
+
+            // Cast to int, will force non-int strings to 0, so we only need to care about negative time as invalid.
+            $valid &= ((int) $runtime >= 0);
+            $valid &= ((int) $startdelay >= 0);
+            $valid &= ((int) $completiondelay >= 0);
+
+            if (!$valid) {
+                return new result(result::ERROR, get_string('taskconfigbad', 'tool_heartbeat', $taskclass));
+            }
+
+            // Lets see if the task is currently running. We can only do this via lockstats.
+            if ($lockstats) {
+                $locked = $DB->record_exists('tool_lockstats_locks', ['resourcekey' => $taskclass]);
+            } else {
+                $locked = false;
+            }
+
+            // Now we can start checking all of the data against the task itself.
+            // First check for start time drift.
+            $starttime = $task->get_next_run_time();
+            if ($startdelay != 0 && ($starttime <= (time() - $startdelay * MINSECS))) {
+                // Lets see if the task is currently running. We can only do this via lockstats.
+                if (!$locked) {
+                    $status = result::CRITICAL;
+                    $messages .= get_string('latencydelayedstart', 'tool_heartbeat', ['task' => $taskclass, 'mins' => $startdelay]) . PHP_EOL;
+                }
+            }
+
+            // Now check for large delays in the last completion.
+            $lastrun = $task->get_last_run_time();
+            if ($lastrun != 0 && $lastrun <= (time() - $completiondelay * MINSECS)) {
+                // Again check if we are locked here. If not, something deeper is wrong, cron might be cooked.
+                // If we ARE locked, we will alert for that instead in the next block.
+                if (!$locked) {
+                    $status = result::CRITICAL;
+                    $messages .= get_string('latencynotrun', 'tool_heartbeat', ['task' => $taskclass, 'mins' => $startdelay]) . PHP_EOL;
+                }
+            }
+
+            // Now we should check the duration of the last run, whenever it was.
+            $dbman = $DB->get_manager();
+            $table = new \xmldb_table('task_log');
+            if ($runtime != 0) {
+                if ($dbman->table_exists($table)) {
```

Comment:
Similarly for this if statement and the elseif statement, it can be moved out and returned earlier as well. The comment would then come before the conditional and be more relevant.